README
README file for pyfred server & clients
---------------------------------------

Content:
    1. Introduction
    2. Servers' and clients' description
       2.1 Filemanager
       2.2 Genzone
           2.2.1 Configuration of genzone client
       2.3 Mailer
       2.4 Techcheck
    3. Repository overview
    4. Configuration
    5. Dependencies
    6. Testing


1. Introduction
---------------

Pyfred is part of FRED system (system for management of internet domains)
and includes various servers and clients both written in python.
In fact there is just one general server which encapsulates specialized
servers, which are often referred to as modules. It is a plugin architecture.
The underlying communication infrastructure is CORBA. Pyfred clients
communicate with servers by RMI paradigm, the interface is specified in
IDL files, which are not part of pyfred's repository, but are centrally
managed elsewhere.


2. Servers' and clients' description
------------------------------------

There are following servers each of which has at least one client:

 i--------------------t-----------------------------------------i
 |        SERVER      |                     CLIENTS             |
 |==============================================================|
 |  filemanager       |          client and admin               |
 |--------------------t-----------------------------------------|
 |  genzone           |          client and test                |
 |--------------------t-----------------------------------------|
 |  mailer            |          client and admin               |
 |--------------------t-----------------------------------------|
 |  techcheck         |          client and admin               |
 |--------------------t-----------------------------------------|

Note: Content of clients column needs some explanation. Almost every interface
      defines two kinds of functions: custom and administrative. Custom
      functions fulfil the purpose of server's existence, it means that
      without these functions there is no reason for server to exist.
      Adminstrative functions constitute completion to normal operation of
      server. They implement more advanced functionality of server. These
      two groups of functions are separated in different files (in different
      clients). Genzone has test client, which is a script intended to be
      used by nagios monitoring software.


 2.1 Filemanager

FileManager daemon is capable of storing and loading files. FileManager
stores files on filesystem under directory which is given in configuration.
FileManager does actually more than just storing or loading file. It keeps
some other information about managed files in database. Currently it is:

  * Numeric id of file
  * Date of creation of record (file's upload date)
  * Human readable and friendly name of file
  * MIME type of file
  * Size of file in bytes
  * Path to file on filesystem in repository
  * Type of file (one of: invoice pdf, invoice xml, accounting xml,
    banking statement)

This set of information can be retrieved by a corba method call and by
another one can be downloaded the file from server. There is also a function
for upload of a file to server. The files are downloaded and uploaded
in client selected chunks in order not to exceed the data size limit
of one corba call.

It is possible to search files based on various criteria, but true
administration interface capable of deletion and manipulation with files
is not yet in place and is not planned for the near future.

Information about files is kept in database compared to files' content, which
is kept on filesystem. The files on filesystem are not stored in flat manner,
but are distributed in subdirectories in order not to overfill one directory.
The relative path is created as:

     {year}/{month}/{day}/{id},

where year, month and day are deduced from date when the file was uploaded
to server. The name of the file on filesystem, here denoted as id, is
primary key of file's record in database and should not be confused with
alias (human-friendly name of file) which is stored as attribute of file
in database and may not be unique.

There are two clients available for FileManager server. One is capable of
download/upload and the second (administrative) of searching in database
of files.


 2.2 Genzone

Goal of the zone generator server is to send all data needed for zone
file creation to a client. It is a task of client to format data in a way,
the DNS server understands to. The tasks of server are:

         *) Generate data which should be in a zone based on status flag
            'outzone'.
         *) Send resulting data through CORBA interface to client in several
            chunks by how big they are.

Zone generator utilizes results of another process, which sets status' for
objects in database. The zone generator simply picks domains, which don't have
status set to 'outzone'.

These constrains must be met in order for domain to be placed in zone. The list
of conditions here is not complete, since the zonegenerator is not responsible
for checking of these conditions.

         *) Domain must have associated nsset.
         *) Domain must not be expired and additionaly if it is an enum domain,
        validation must be still valid. By 'expired' we mean that
        domain is out of safe-period (which is about one month after the
        real expiration date).
         *) The domain must not have set status 'outZoneManual', it means
            the domain was not manually outaged from zone.

A nameserver which requires GLUE by DNS standard is considered valid
only if it has one. On the other side, if nameserver has a GLUE and the GLUE
is not required, the GLUE is not passed to client but the nameserver is.
Neither of mentioned inconsistencies influences placement or displacement
of a domain in a zone. Only warning messages are logged to syslog.

The client is responsible for formatting of received data in manner which is
acceptable by DNS server.  The only currently availabel output format is a
format of BIND DNS server. Second client script is dedicated solely to testing
of genzone server. Zone generation is so important service that there is a
special need to test, that server is up and running. Output of test script is
compatible with expectations of Nagios monitoring software.  Genzone client
handles also backups of old zone files before they are overwritten by newly
generated zone files. It is possible to configure post-hook for each zone (or
in DEFAULT section for all zones).  Hook is a command which is executed after
the zone is generated.  Genzone performs substitution for following tokens in
hook command line:

  $file     absolute path of new generated zone file
  $zone     name of zone which is generated
  $backup   absolute path of previous (backed up) zone file

The hook is trigered only if the zone file is successfully generated. By this
hook you can conveniently restart a bind server for example.

Zone generation support generation of DS records from server keyset objects
attached to domains.

  2.2.1 Configuration of genzone client

Genzone client is configurable by configuration file. The default location
is /etc/fred/genzone.conf, but you may override this default from command
line. The configuration file may contain individual section for each zone.
Configuration parameters, which are not specific to a zone, are in section
called 'general'. The command line parameters are preferred to values
from configuration file, but beware that options on command line override
configuration of all generated zones. The description of configuration
directives follow.

  General configuration directives:

    *) nameservice [localhost] - Host where is running corba nameservice.

    *) context [fred] - Corba nameservice context (directory)

    *) verbose [Off] - If the client should be verbose when generating zones.

    *) zones [] - Space separated list of zone names to be generated. This
       list is only used, if no zones were specified on command line. If no
       zones are specified at all, list of zones is downloaded from server

  Zone specific configuration directives:

    *) zonedir [./] - Directory where new zone files should be generated.
       Default is current directory.

    *) backupdir [] - Directory where current zone file from zonedir should
       be backuped before being overwritten by new zone file. Default is
       to use the same directory as zonedir.

    *) nobackups [False] - It set to True, no backups are created at all
       (backupdir directive is in that case useless).

    *) chunk [100] - Relates to mechanism of how the records are transfered
       over corba. It means how many domains are transferred in one CORBA
       call.

    *) header [] - If defined, the content of header file will be prepended
       to zone file.

    *) footer [] - If defined, the content of footer file will be appended
       to zone file.

    *) maxchanges [-1] - The maximal number of changed records between the
       new zone file and previous zone file. If The number is negative, it
       means that the is no such a limit. If you prepend '%' character to
       the number, it means procentual change.

 2.3 Mailer

Mailer daemon is responsible for delivering and archiving email messages.
It is a CORBA server with interface consisting of two parts:

    1. User interface is one function for sending of emails.
    2. Admin interface is a set of functions for mailarchive searching.

Part of configuration is stored in database and part in a configuration file.
Sending of email consists of following steps:

    1) Merging client provided mail headers and defaults stored in database.

    2) Running dataset provided by client merged with dataset from
       database through templates and subject, both selected according to
       email type. Mail type may contain multiple templates, each template will
       produce one part of multipart email message of MIME type text/*.

    3) So far generated email is archived in database in text form together
       with some meta information (creation date, associated handles, ...).
       The email is marked as ready to be sent.

Then there is another process which periodically activated and it does:

    1) The email, which is ready to be sent, is read from database.

    2) Attachments not intended to be templated are retrieved from file-manager
       daemon, base64-encoded and attached to email message.

    3) The whole MIME multipart email message is transformed to text and this
       text is signed. New SMIME email is created.

    4) The email is sent by sendmail program. If sendmail is successfull,
       the email in archive is marked as sent otherwise number of attempts
       is incremented.

If email cannot be sent three times in row (configurable), sending of email is
given up.  The signing is done by openssl binary. M2Crypto python library
cannot be used, allthough it would be better solution, because emails signed by
m2crypto cannot be validated in MS Outlook.

Searching of mail archive for an email is based on "filter" and method of
incremental transfers. Filter is a structure which specifies constraints,
which must be fulfilled by email message in order to be chosen for retrieval.
The set of results is then transfered in chunks, of which size can be specified
by client. When all data are transfered or they are not wanted anymore, the
transfer should be closed.

As template system is used clearsilver, which is shared library, which can
be used from C and Python. Clearsilver templates are considerably easier
to use than xslt and at the same time have equal expression capabilities.
Information how to write clearsilver templates can be found at:
http://www.clearsilver.net/docs/man_templates.hdf. We will give just a few
examples, which should be sufficient for 99% of email templates, since
email templates aren't expected to be much complicated.

All substitution commands are enclosed in '<?cs' and '?>'.

  <?cs var:foo ?>    This will output variable 'foo' if it is defined.

  <?cs if:bar ?><?cs var:bar ?><?cs else ?>There is nothing to see<?cs /if ?>
  This is example of condition. If variable 'bar' is True then content of
  it is displayed. Otherwise the text after else is displayed.

The values for variables (in our case 'foo' and 'bar') are provided by
parameter of function for sending email, which is a list of key-value
structures.

Iterations can be done in following way (assuming we have defined myset,
which contains elements, which in turn have attributes key and value):

  <?cs each:item = myset ?>
    Key of element of myset: <?cs var:item.key ?>
    Value of element of myset: <?cs var:item.value ?>
  <?cs /each ?>

There exists a utility which makes working with templates easier. It is the
script template_sucker, which is capable of printing templates stored in
database in well arranged manner and printing of variables used in individual
templates.

As ussually there exist two client utilities for mailer. One is for email
submission and another one is for searching email archive.

Mailer has capability of checking whether an email was delivered successfully
under assumption that remote smtp server sends error message to "envelope
from" address, if something goes wrong. The smtp server responsible for
delivery of emails for domain used in "envelope from" must be aware of
pyfred and deliver these error messages to special mailbox, which is
accessed by mailer module in regular intervals and checked for new emails.
If user part of email address in "To" header matches identifier associated
with sent email archived in database, the email is marked as undelivered.
This process of downloading emails over imap is optional and may be turned
off in configuration file.

Emails in archive may have assigned several status values, which are numbers
from 0 to 4. Here is their meaning:

    0: Mail was successfully sent.
    1: Mail is ready to be sent.
    2: Mail waits for manual confirmation.
    3: This email will not be sent or touched by mailer.
    4: Delivery of email failed.


 2.4 Techcheck

Techcheck daemon is responsible for testing of nameservers in nsset. There
are two possible ways how to call out technical check.

  1) You can do technical check on a single nsset synchronously. It means
  that the caller is blocked until the tests are done and then receives
  results in form of return value from CORBA function.

  2) Another way is to do the test of nsset asynchronously. It means
  that server returns from CORBA function as soon as it has found all
  relevant data needed for technical check in database. The registrator
  is informed about results of technical tests in form of a poll message
  over EPP interface.

The technical check consists of several tests. The tests are highly dynamic and
are not hardcoded in source code (except one). The possible tests are defined
in database and when the pyfred starts up, they are read and so called "test
suite" is constructed, which is used in all checks untill the pyfred is
restarted. Test suite is a sequence of technical tests which should be applied
if they are not disabled by certain flag in database. There are a few
attributes of test, the most important is:

  * numeric id of test (influences order of test in testsuite)
  * name of test
  * level of test (this is how the test is detailed, 1..10)
  * information about dependency on another test(s)
  * a path to script which executes the testing
  * flag saying if the test is meaningfull without domain fqdn(s)

Basic schema of technical check is as follows:

  1) test is popped from test suite and if the dependencies are satisfied
  and its level is lower or equal to specified level, the test is executed.
  There is one more exception when the test is not executed. It is when the
  test requires a list (possibly containing only one item) of domain fqdns in
  order to be meaningfull, but there are not any fqdns.

  2) test is realized by execution of appropriate test script. Robustness
  in respect to potencial script freeze-up is accented. As last option
  is used SIGKILL to end life of child.

  3) from return value of the script is deduced status of the test. Three
  values are possible: passed, failed, unknown. Unknown denotes that
  the script failed because of unexpected (unknown) reason or the reason
  is known, but it is not in scope of that particular test.

  4) poll message is generated for registrar if the check was submitted
  over EPP interface (asynchronous way) or the caller is informed
  about the results through return value of CORBA function (synchronous
  way).

  5) individual results of tests are archived in database.

There are two basic categories of tests. Tests which require list of domain
fqdns and tests which test nameservers as such without need of specific
domain fqdns.

There is also interface accomodating search in executed technical checks.
The interface is pretty similar to search interfaces in other modules ("filter
and search object" model).

In background are continuosly performed technical checks of all registered
nssets. This process is called regular (or periodic) technical checks.
These can be turned of in config file.

There is one special test, which tests necessity of GLUE for nameserver. This
test must be executed as first one, because it transforms a list of ip address
and all other tests use this modified list. All other tests should depend on
this test. This test is exceptional also because it is not executed by external
script but directly in techcheck, though it has a record in database, which
has empty string in 'script' field.


3. Repository overview

Here we describe what is in which directory and where it is installed.

  pyfred directory - is a python package which contains:

    * 'zone' module used by genzone client for communication with CORBA server.
    * 'utils' module which gathers various functions shared by corba servers.
    * 'modules' package which contains modules implementing the CORBA servers.

  scripts directory - contains python scripts (pyfred server and all clients)
    which are installed in $PATH. There is also start/stop/status pyfredctl
    script.

  tc_scripts directory - contains scripts, currently only python scripts but
    there may be any other scripts, which are used by techcheck server to
    test a nameserver. These scripts are not installed in $PATH because they
    are not ment to be run from command line, but rather in
    ${PREFIX}/libexec/pyfred.

  misc directory - contains everything else what didn't fit in any of previous
    directories. Currently there is template_sucker.py script, which ease
    debugging of mailer templates, and directory mailer_examples contains
    prepared data files, which may be used together with mailer client.


4. Configuration

The pyfred server looks for its configuration file (pyfred.conf) in current
directory, /etc/fred or /usr/local/etc/fred in this order. When starting it
prints, which config file it is using. You can explicitly set configuration
file on command line.

Configuration file is organized in sections. One called General which is not
specific to any CORBA server and influences general behaviour shared by
all CORBA servers, the other sections are dedicated to individual CORBA
servers.

Before we will look at configuration directives in depth, we will explain
meaning of 2 directives which are common for all CORBA servers (it means
they can be set in all sections except section General). It is 'idletreshold'
and 'checkperiod' used to control lifetime of dynamically created CORBA
objects, both accept number of seconds. Every CORBA server has by coincidence
need to spawn new CORBA objects, which are dedicated to one simple task
(i.e. to transfer bytes of file or to transfer a zone records).

 *) idletreshold [3600] - says how many seconds the object must not be
 accessed by any CORBA method in order to be taken as "death" and will be
 automically deleted.

 *) checkperiod [60] - is interval in which are checked objects for being idle.

The following rule should hold: checkperiod <= idletreshold.

The sections will be examined in following order:

  1) General
  2) Genzone
  3) Mailer
  4) FileManager
  5) TechCheck

The default values for directive is enclosed in square brackets. If there is
no default value, the square brackets are missing.

  General
  -------

  *) modules - list of modules (CORBA servers) which should be loaded and
     started.

  *) dbhost - Database host (empty value means: use unix socket).

  *) dbport [5432] - Database port (5432 is standard postgresql's port).

  *) dbuser [fred] - Database username.

  *) dbpassword - Database password.

  *) dbname [fred] - Database name to use.

  *) nshost [localhost] - Host where CORBA nameservice runs.


ChangeLog
2020-01-31 (2.13.1)

    * Fix rpm for RHEL8 and F31

2019-11-26 (2.13.0)

    * Use setuptools for distribution

2019-06-27 (2.12.1)

    * Fix rpm build

2019-03-18 (2.12.0)

    * License GNU GPLv3+

2019-02-11 (2.11.1)

    * Add systemd service for fedora package

2018-11-20 (2.11.0)

    * COPR build

    * AUTHORS file update

2018-01-16 (2.9.0)

    * Static IDL for python (precompiled)

2017-03-02 (2.8.0)

    * Configuration files documentation

    * Fedora packaging

2017-03-30 Jiri Sadek (2.7.3)

    * replace usage of user-defined aggregate function array_accum with built-in array_agg

2016-03-21 Jaromir Talir (2.7.2)

    * Fix rpm build

    * Removed unused idl references in setup.py

2015-08-26 Jiri Sadek (2.7.1)

    * fix techcheck - glue test

    * fix techcheck scripts - enable edns (tc responses)

    * genzone - support for sha256 digest algorithm for ds records

2015-01-27 Jan Zima (2.7.0)

    * domain browser code removed

    * techcheck - regular check query optimization

    * mailer - add support for idn

2015-01-13 Michal Strnad (2.6.2)

    * fix zone generator - ip address type resolution

2014-08-01 Jan Zima, Zdenek Bohm, Jaromir Talir (2.6.1)

    * removed domain browser from default modules

    * m2crypto dependency fixes

2014-02-13 Michal Strnad (2.6.0)

    * Mail default header is now controlled by mail type

    * Support for multiple signing certificates (distinguished by sender e-mail address)

2014-02-17 Jaromir Talir (2.5.3)

    * Fix rpm dependencies

2013-11-06 Jaromir Talir, Zdenek Bohm (2.5.2)

    * Fix genzone - glue records generation

    * Fix domain browser
        - date time format fix reverted (fixed in web app)
        - registrar address formating
        - domain detail public/private data recognition (handle -> id fix)

2013-09-12 Jiri Sadek (2.5.1)

    * Fix logging of job name (job_noop issue)

2013-08-14 Zdeněk Böhm (2.5.0)

    * domain browser - implementation of idl interface

2013-07-29 Jaromir Talir (2.4.4)

    * Fix fedora rpm package build

2013-06-27 Jiri Sadek (2.4.3)

    * Removed hardcoded mail type priorities - now loaded from database

2013-04-24 Jiri Sadek (2.4.2)

    * Temporary enhancement for email sending - hardcoded mail type priorities

2012-10-18 Vlastimil Zima (2.4.1)

    * Fix installation of additional directories

2012-09-06 Jiri Sadek, Tomas Divis, Vlastimil Zima (2.4.0)

    * Whitespace normalization and PEP8-ification

    * Update due to distutils changes (setup.cfg)

    * Mailer - email type penalization code simplified

    * Optional PID file

    * Configure option for number of wait round for external mail signing command

2012-07-13 Jiri Sadek (2.3.1)

    * add more debug logging to pyfred main loop

2012-05-14 Jiri Sadek (2.3.0)

    * Mailer - add support for IMAP SSL

2011-06-02 Jiri Sadek (2.2.2)

    * Mailer - little tweak to not loose one send period because of penalized
               mail types when no other mail types are ready to send

2011-05-30 Jiri Sadek (2.2.1)

    * Mailer - bugfixed list of attachments

2011-05-23 Jiri Sadek (2.2.0)

    * Mailer - new approach for selecting emails ready to send
             - fixed connection handling in mailNotify method

    * fixing dnssec technical test - new version of ldns drill utility

2009-11-09 Jiri Sadek (2.1.9)

    * Fixed configuration defaults (logging option)

    * Technical tests now notify only newest state (history) of nsset
      (parallel nsset of different states fix)

    * Fixed email signing procedure - exception type changed

2009-07-09 Jiri Sadek, Tomas Divis (2.1.8)

    * Fixes in mail module to properly update number of attempts (sending
      mails)

    * External processes is now executed with common 'runCommand' method with
      timeout settings

    * Fixes in installation procedure, implemented --no-check-deps for
      disabling of checking dependencies

    * Removed use of _quote() method from pygresql library - due to changes in
      pygresql-4.0

    * Logging system rewritten to support different handlers -
      syslog/file/console (using logging python module)

    * Bugfix in configuration template

2009-07-03 Jiri Sadek (2.1.7)
    * Bugfixes for previous release (last was quite broken)
        - install procedure
        - dnssec test - more error handling

2009-06-24 Jiri Sadek (2.1.6)

    * Test for dnssec key chain of trust added. Test uses drill utility (this
    add new dependency to project)
    Test copy current approach, but changes was needed:
        - __dbGetAssocDomains(...) now return dictionary where key is fqdn of
        domain, value is True/False flag defining if there is a keyset
        associated with domain (we need test only this domains)
        - in database table `check_test' new value (3) is used in need_domain
        column determining that test needs only signed domains on standard
        input

    * New configuration options:
        - drill binary executable
        - trust anchor key file

2009-03-14 Tomas Divis (2.1.5)

    * Bugfix in sending emails from tech check module - it generated invalid
      corba request because of empty list of email addresses.

2009-03-26 Tomas Divis (2.1.4)

    * When marking emails as undelivered, response is saved using base64
      encoding (due to SQL errors when non-utf8 response was delivered).
      Old data must be migrated:
      UPDATE mail_archive SET response = encode(convert_to(response, 'UTF8'),
      'base64')

2009-02-10 Tomas Divis (2.1.3)

    * Adding few log messages to debug memory consumtion

2008-11-10 Ales Dolezal (2.1.2)

    * Little fix in installation procedure
    - MANIFEST.in updated

2008-11-08 Jaromir Talir (2.1.1)

    * Renaming
    - pyfred_server -> fred-pyfred
    - genzone_test -> check_pyfred_genzone

2008-10-19 Jaromir Talir (2.1.0)

    * Adding DS generation from DNSKEY records

2008-09-18 Jaromir Talir (2.0.1)

    * Fixing zone generator
    - syntax error

server/pyfred.py
#!/usr/bin/env python
# vim:set ts=4 sw=4:
"""
This module is a framework in which can be easily embedded various corba
object implementations through extending modules defined in modules.conf.
Those objects are registered in corba name service and the rest is up to
a caller (corba client).
"""

import sys, time, syslog, ConfigParser
from omniORB import CORBA, PortableServer
import CosNaming
import pgdb

def job_noop():
	"""
This function is used as placeholder in job queue if the queue is empty.
	"""
	pass

class Logger(object):
	"""
This class encapsulates standard syslog function. The only difference is
that each message when called through Logger is prepended by name of module,
which is set during Logger's initialization.
	"""
	EMERG = syslog.LOG_EMERG
	ALERT = syslog.LOG_ALERT
	CRIT = syslog.LOG_CRIT
	ERR = syslog.LOG_ERR
	WARNING = syslog.LOG_WARNING
	NOTICE = syslog.LOG_NOTICE
	INFO = syslog.LOG_INFO
	DEBUG = syslog.LOG_DEBUG

	def __init__(self, prefix):
		"""
	Initialize the prefix which will be used for every message logged
	trhough this Logger.
		"""
		self.prefix = prefix

	def log(self, level, msg):
		"""
	Wrapper around syslog.syslog() function which adds prefix to every
	logged message.
		"""
		syslog.syslog(level, "%s module - %s" % (self.prefix, msg))

class DB(object):
	"""
This class provides methods usefull when working with database.
	"""

	def __init__(self, dbhost, dbport, dbname, dbuser, dbpassword):
		"""
	Method initializes data needed for database access.
		"""
		self.host = dbhost
		self.port = dbport
		self.dbname = dbname
		self.user = dbuser
		self.password = dbpassword

	def getConn(self):
		"""
	Obtain connection to database.
		"""
		return pgdb.connect(host = self.host +":"+ self.port,
				database = self.dbname, user = self.user,
				password = self.password)

	def releaseConn(self, conn):
		"""
	Release connection obtained in getConn() method.
		"""
		conn.close()

def getConfiguration(configs):
	"""
Get configuration from file. The configuration files are given as a list in
parameter. The function automatically provides default values for directives
not defined by configuration file. If none of the configuration files exists,
all directives will contain default values.
	"""
	# set defaults
	confparser = ConfigParser.SafeConfigParser({
			"context":"fred",
			"dbhost":"",
			"dbname":"ccreg",
			"dbport":"5432",
			"dbuser":"postgres",
			"dbpassword":"",
			"nshost":"localhost",
			"nsport":"2809",
			"loglevel":"LOG_INFO",
			"logfacility":"LOG_LOCAL1",
			"port":"2225"})
	# read configuration file
	for cfile in configs:
		try:
			if len(confparser.read(cfile)) == 1:
				print "File %s used as config file" % cfile
				break
		except ConfigParser.MissingSectionHeaderError, e:
			sys.stderr.write("Error in configuration file '%s': %s\n" %
					(cfile, e))
			return None
		except ConfigParser.ParsingError, e:
			sys.stderr.write("Error in configuration file '%s': %s\n" %
					(cfile, e))
			return None
	# create basic section general if it does not exist
	if not confparser.has_section("General"):
		confparser.add_section("General")
	return confparser

#
# main
#
def main(argv):
	# if server should detach from terminal after startup
	detach = True
	# default places where to look for configs
	configs = ["/etc/pyfred.conf", "/usr/local/etc/pyfred.conf",
	           "pyfred.conf"]
	for arg in argv[1:]:
		if arg == "-d":
			detach = False
		else:
			# append explicit config location if present on command line
			configs.insert(0, arg)
	# get configuration
	conf = getConfiguration(configs)
	if not conf:
		sys.exit(1)

	# open syslog (syslog function will be later passed to modules)
	try:
		logfacility = syslog.__dict__[ conf.get("General", "logfacility").upper() ]
	except KeyError, e:
		sys.stderr.write("syslog facility is invalid: %s\n" %
				conf.get("General", "logfacility"))
		sys.exit(1)
	syslog.openlog("pyfred", (syslog.LOG_PID | syslog.LOG_CONS), logfacility)
	try:
		loglevel = syslog.__dict__[ conf.get("General", "loglevel").upper() ]
	except KeyError, e:
		sys.stderr.write("log level is invalid: %s\n" %
				conf.get("General", "loglevel"))
		sys.exit(1)
	syslog.setlogmask(syslog.LOG_UPTO(loglevel))

	# create database object to be used in modules
	db = DB(conf.get("General", "dbhost"),
			conf.get("General", "dbport"),
			conf.get("General", "dbname"),
			conf.get("General", "dbuser"),
			conf.get("General", "dbpassword"))
	# get a list of modules to import
	modulenames = conf.get("General", "modules").split(" ")
	# update import path
	sys.path.insert(0, "idl")
	sys.path.insert(0, "/usr/lib/pyfred/share")
	sys.path.insert(0, "/usr/lib/pyfred/server")
	# load all modules
	modules = []
	for mname in modulenames:
		try:
			print "Importing module %s" % mname,
			modules.append(__import__(mname))
			print " ... ok"
		except Exception, e:
			print " ... failed"
			sys.stderr.write("Could not import module '%s'.\n" % mname)
			sys.stderr.write("Python exception: %s\n" % e)
			sys.exit(1)

	try:
		# Initialise the ORB and find the root POA
		nsname = "NameService=corbaname::" + conf.get("General", "nshost") + \
				":" + conf.get("General", "nsport")
		orb = CORBA.ORB_init(["-ORBnativeCharCodeSet", "UTF-8", "-ORBInitRef",
				nsname, "-ORBendPoint", ":::" + conf.get("General", "port")],
				CORBA.ORB_ID)
		rootpoa = orb.resolve_initial_references("RootPOA")
		# create persistent poa
		ps = [rootpoa.create_lifespan_policy(PortableServer.PERSISTENT),
			  rootpoa.create_id_assignment_policy(PortableServer.USER_ID),
		      rootpoa.create_implicit_activation_policy(PortableServer.NO_IMPLICIT_ACTIVATION)]
		poa = rootpoa.create_POA("pyfredPOA", rootpoa._get_the_POAManager(), ps)

		# Obtain a reference to the root naming context
		obj	= orb.resolve_initial_references("NameService")
		rootContext = obj._narrow(CosNaming.NamingContext)
		if rootContext is None:
			sys.stderr.write("Failed to narrow the root naming context")
			sys.exit(1)
	except CORBA.TRANSIENT, e:
		sys.stderr.write("Is nameservice running?\n(%s)" % e)
		sys.exit(1)
	except CORBA.Exception, e:
		sys.stderr.write("CORBA failure, original exception is:\n%s" % e)
		sys.exit(1)

	# Bind a context named "fred.context" to the root context
	# This context is a container for all registered objects
	name = [CosNaming.NameComponent(conf.get("General", "context"), "context")]
	try:
		fredContext = rootContext.bind_new_context(name)
		print "New fred context bound"
	except CosNaming.NamingContext.AlreadyBound, ex:
		print "fred context already exists"
		obj = rootContext.resolve(name)
		fredContext = obj._narrow(CosNaming.NamingContext)
		if fredContext is None:
			print "fred.context exists but is not a NamingContext"
			sys.exit(1)

	# Run init methods in all registered modules and bind their objects
	# to fred context
	joblist = []
	for module in modules:
		try:
			servant, name = module.init(Logger(module.__name__), db, rootContext,
					conf, joblist, rootpoa)
		except Exception, e:
			sys.stderr.write("Error when initializing module '%s': %s:%s\n" %
					(module.__name__, sys.exc_info()[0], e))
			sys.exit(1)
		poa.activate_object_with_id(name, servant)
		ref = poa.id_to_reference(name)
		cosname = [CosNaming.NameComponent(name, "Object")]
		try:
			fredContext.bind(cosname, ref)
			print "New '%s' object bound" % name
		except CosNaming.NamingContext.AlreadyBound:
			fredContext.rebind(cosname, ref)
			print "Existing '%s' object rebound" % name

	# Activate objects
	# Activate the POA
	poaManager = poa._get_the_POAManager()
	poaManager.activate()

	# redirect stdin, stdout, stderr to /dev/null. Since now we will use 
	# only syslog
	if detach:
		print "Detaching from terminal"
		sys.stdin.close()
		sys.stdout.close()
		sys.stderr.close()
		sys.stdin  = open("/dev/null", "r")
		sys.stdout = open("/dev/null", "w")
		sys.stderr = open("/dev/null", "w")
	print "Logging to syslog since now"
	syslog.syslog(syslog.LOG_NOTICE, "Python fred Server started.")
	syslog.syslog(syslog.LOG_INFO, "Loaded modules: %s" % modulenames)
	# Run cron jobs forever
	if len(joblist) == 0:
		joblist = [ { "callback":job_noop, "context":None, "period":3600 } ]
	delay = 5 # initial delay
	while True:
		time.sleep(delay)
		# look what is scheduled for execution
		job = joblist.pop(0)
		# execute first job in a qeueu
		try:
			job["callback"](job["context"])
		except Exception, e:
			syslog.syslog(syslog.LOG_ERR, "Unexpected error when "
					"executing job. %s:%s" % (sys.exc_info()[0], e))
		# schedule job for next execution
		job["ticks"] = job["period"]
		for i in range(len(joblist)):
			item = joblist[i]
			if job["ticks"] - item["ticks"] >= 0:
				job["ticks"] -= item["ticks"]
			else:
				joblist.insert(i, job)
				job = None
				break
		if job:
			# job should be run as last
			joblist.insert(len(joblist), job)
		# calculate the time to next awakening
		delay = joblist[0]["ticks"]


	# Block for ever (or until the ORB is shut down)
	#orb.run()

if __name__ == "__main__":
	main(sys.argv)
